### ISSUE TITLE
Memory usage ever grows - data leak

### ISSUE THREAD

**Issue by r4f4ss** [optimism-java/shisui#215]
EDIT: Execution with parameter (3 networks) `--netrorks "history,beacon,state"`

The figure below is the dashboard, where one can see in the right-top corner the graph showing the memory that ever growing.
![grafana9](https://github.com/user-attachments/assets/6a453752-f92e-42f4-8587-4b2baff9d7fd)

The same happens with metrics disables, the memory usage can be viewed with `top` and show the same behavior:
![memory1 1](https://github.com/user-attachments/assets/593aa0a3-90e1-49ca-be73-9cfe587eb1ed)
![memory1 2](https://github.com/user-attachments/assets/1d46ab7a-edfe-4d12-9549-355fe7e96e66)
![memory1 3](https://github.com/user-attachments/assets/23307e1f-a017-4964-ba7b-dadc4e7a3fbf)

Eventually the PC freeze by lack of memory.

The cause is related with github.com/optimism-java/utp-go/buffers.NewSyncBuffer, as shown by `pprof`: 
![image](https://github.com/user-attachments/assets/4b0c7c52-2114-4e96-bb27-656176c88e50)

**Comment by GrapeBaBa** [comment]
@thinkAfCod  It seems that the memory leak issue has reappeared

**Comment by r4f4ss** [comment]
Thanks for the explanation @GrapeBaBa , I believe you are correct.

I managed to reduce the leak by:

1. Setting a self delete after 180 seconds of `t.activeCallByAuth[nonce]` [here](https://github.com/optimism-java/shisui/blob/portal/p2p/discover/v5_udp.go#L618-L619)
2. Setting a self close after 20 seconds of created `utpConn` at utp-go lib [here](https://github.com/optimism-java/utp-go/blob/295a86339e8b497decd1d984ecb86613fa455e1d/utpgo.go#L1154)
2. Setting a self close after 20 seconds of created `utpConn` droped of backlog (full) [here](https://github.com/optimism-java/utp-go/blob/295a86339e8b497decd1d984ecb86613fa455e1d/utpgo.go#L1179)

It is not a solution since it breake the properly workig of shisui, but serves to prove places where the leak happens. There is still a slow memory usage grow, not solved yet.

![image](https://github.com/user-attachments/assets/b0e7cb9c-94e3-4382-86bc-7b3e61aafe49)

![image](https://github.com/user-attachments/assets/e2e83b89-a75d-48cb-bb8a-da0e2330e059)


Looking forward to see @thinkAfCod  approach.

**Comment by GrapeBaBa** [comment]
> I'm trying this solution and will know by tomorrow if it works: #219 .
> 
> And, how can I reproduce this more quickly? I currently have 10 nodes running on the server, but their memory growth is very slow. It's not like what you showed, where the memory can grow to around 4GB in a hour.

We could setup influxdb+grafana locally

**Comment by r4f4ss** [comment]
> And, how can I reproduce this more quickly? I currently have 10 nodes running on the server, but their memory growth is very slow. It's not like what you showed, where the memory can grow to around 4GB in a hour.

I have been using locally pprof + [grafana and influxdb](https://github.com/optimism-java/shisui/blob/portal/grafana/README.md)

As a qick rule, the plot of memory in grafana should become constant under 1 hour, but a day test may be more precise becuse it deal with high and low trafic periods. Basically, if the memory only grows, even after the garbage collector pattern in the plot (it is like a triangle), there is some leak.

**Comment by thinkAfCod** [comment]
> @thinkAfCod I really can not explain why my measures are diferent from the memory usage you have in your server, I will let it runnig for several hours (with and without grafana), at least, to have a more conclusive answer.

Don't worry, this PR( #219 ) has no effect. The memory leak occurs in the dispatch method of Udp_v5, where some cached packets are not being cleared. I need to think about how to clean up these caches.

**Comment by thinkAfCod** [comment]
in heap of pprof, this part's memory usage keeps increasing:
<img width="956" alt="image" src="https://github.com/user-attachments/assets/d60a8cee-5266-4b0b-825f-ff0d7e93bca7">

**Comment by r4f4ss** [comment]
Just noted that runnig the three subnets (`--networks "history,beacon,state"`) increase considerably the memory growth , I guess this bug related to execution with three networks. 

I confirm that executing with default configuration (only history) gives me a slow memory growth and a pprof reported above by @thinkAfCod .

**Comment by thinkAfCod** [comment]
The memory leak is likely occurring at the point where new UTP connections are accepted. 

There should be a large number of connections timing out here, but they are not being cleaned up. A cleanup task needs to be added.

```
func (l *Listener) AcceptUTPContext(ctx context.Context, connIdSeed uint32) (*Conn, error) {
	l.recordRequireConnId(connIdSeed)
	ticker := time.NewTicker(time.Millisecond * 17)
	defer ticker.Stop()
	for {
		select {
		case newConn, ok := <-l.acceptChan:
			if ok {
				if conn := l.filterConn(newConn, connIdSeed); conn != nil {
					return conn, nil
				}
				continue
			}
			err := l.encounteredError
			if err == nil {
				err = l.makeOpError("accept", net.ErrClosed)
			}
			return nil, err
		case <-ticker.C:
			if conn := l.removeAndGetCommingConn(connIdSeed); conn != nil {
				return conn, nil
			}
		case <-ctx.Done():
			return nil, ctx.Err()
		}
	}
}
```

**Comment by r4f4ss** [comment]
I tried to close every connection (`newConn` and `conn`) after 20 seconds with a goroutine.
After run for about 1:30 hours it does not solves

![fig2](https://github.com/user-attachments/assets/8f4064a6-35ee-4ae2-84cc-3ff32114c873)

![image](https://github.com/user-attachments/assets/648e76fa-8b51-4b17-b0a6-a5c1a1196089)

May be conn is not been closed properly.

**Comment by thinkAfCod** [comment]
Yes, it is necessary to modify the code of the utp-go library.

Simply modifying the code of shisui will not have any effect.

**Comment by thinkAfCod** [comment]
I modified the code of utp-go and started 5 nodes on the server.

These nodes simultaneously launched the `state` network and the `history` network.

I will continue to monitor their memory usage.

They have been running for nearly 40 minutes now, with a maximum memory usage of 4GB.

@r4f4ss Could you help monitor it as well?

**Comment by r4f4ss** [comment]
Thank you @thinkAfCod . How can I help to monitor the nodes?

**Comment by thinkAfCod** [comment]
Just start an empty data node, along with the state and history networks, and then use Grafana to record the maximum memory usage, which can serve as a basic memory usage metric.

**Comment by r4f4ss** [comment]
I started an empty data node running code from PR #219  with command: 
```sh
./build/bin/shisui --nat stun --networks "history,state" --pprof --metrics --metrics.influxdb --metrics.influxdb.endpoint "http://0.0.0.0:8086" --metrics.influxdb.username "shisui" --metrics.influxdb.password "shisui" --metrics.influxdb.database shisui --data.dir data
```

After run for more than 4h, the memory held is about 100 MB and memory used is roughly constant:
![image](https://github.com/user-attachments/assets/d1c21786-6cfa-4248-9f58-d426a9ce4de2)

Profile is not dominated anymore by `buffers.NewSyncBuffer` and memory is spreaded among many structures:
![image](https://github.com/user-attachments/assets/91c825b0-4c4a-4b4d-894e-5ff0366de05c)

Seems very good, I will keep monitoring.

**Comment by r4f4ss** [comment]
fixed in #219
